### PR TITLE
feat: Change middleware to also handle AuthV2

### PR DESCRIPTION
**Pull request by chanioxaris**

### CODE REVIEW

File: src/Clerk.ts [hunk 1]
```diff
@@ -150,25 +152,29 @@ export default class Clerk {
...
     if (!decoded) {
       throw new Error(`Failed to verify token: ${token}`)
     }
 
     const key = await this._jwksClient.getSigningKey(decoded.header.kid)
-    const verified = jwt.verify(token, key.getPublicKey(), { algorithms: algorithms as jwt.Algorithm[], audience: 'clerk'})
+    const verified = jwt.verify(token, key.getPublicKey(), {algorithms: algorithms as jwt.Algorithm[]}) as JwtPayload
 
-    return verified as jwt.JwtPayload
+    if (!verified.hasOwnProperty('iss') || !(verified.iss?.lastIndexOf('clerk.', 0) === 0)) {
```
Comment (agis) [inline src/Clerk.ts:173]: 🔧  This starts with `https://clerk.` - see https://github.com/clerkinc/clerk-sdk-ruby/pull/9/files#diff-b8dd9279fc79620beffbf4ca787c70a26d5fe0208f082a64fee4f16211c7340fR178

Comment (chanioxaris) [reply src/Clerk.ts]: Fixed

File: src/Clerk.ts [hunk 2]
```diff
@@ -150,25 +152,29 @@ export default class Clerk {
...
       throw new Error(`Failed to verify token: ${token}`)
     }
 
     const key = await this._jwksClient.getSigningKey(decoded.header.kid)
-    const verified = jwt.verify(token, key.getPublicKey(), { algorithms: algorithms as jwt.Algorithm[], audience: 'clerk'})
+    const verified = jwt.verify(token, key.getPublicKey(), {algorithms: algorithms as jwt.Algorithm[]}) as JwtPayload
 
-    return verified as jwt.JwtPayload
+    if (!verified.hasOwnProperty('iss') || !(verified.iss?.lastIndexOf('https://clerk.', 0) === 0)) {
+        throw new Error(`Invalid issuer format: ${verified.iss}`)
```
Comment (SokratisVidros) [inline src/Clerk.ts:174]: 🙃 
```suggestion
        throw new Error(`Invalid issuer: ${verified.iss}`)
```

**Review by SokratisVidros — APPROVED** [review]
👏

**Comment by SokratisVidros** [comment]
@chanioxaris Can you please remind me how do we fetch the JWKS key in the middleware? Is it upon initialization? Do we do it once?

**Comment by chanioxaris** [comment]
> @chanioxaris Can you please remind me how do we fetch the JWKS key in the middleware? Is it upon initialization? Do we do it once?

The clerk class contains a `jwksClient` which is initialized within the constructor. There is a built-in memory cache in there, that holds the keys with a default cache max age of 1 hour.

File: src/Clerk.ts [hunk 3]
```diff
@@ -194,6 +200,14 @@ export default class Clerk {
   }
 
   expressWithSession({ onError }: MiddlewareOptions = { onError: this.defaultOnError }): (req: Request, res: Response, next: NextFunction) => Promise<void> {
+    function isAuthV2Request(clerk: Clerk, token: undefined | string): boolean {
+      if (!token) {
+        return false
+      }
+
+      return clerk.decodeToken(token).hasOwnProperty('sid');
```
Comment (agis) [inline src/Clerk.ts:208]: :wrench: In Ruby we decide if it's an authv2 token based on if `iss.starts_with("https://clerk")`. I suggest to do the same here, as it's more explicit than simply checking if `sid` is present (which can be present from 3rd-party tokens as well).

File: src/Clerk.ts [hunk 4]
```diff
@@ -204,36 +218,45 @@ export default class Clerk {
...
-        Logger.debug(`sessionId from query: ${sessionId}`);
-
-        let session: (Session | undefined) = undefined;
-        if (!sessionId) {
-          const client = await this.clients.verifyClient(sessionToken);
-          session = client.sessions.find(session => session.id === client.lastActiveSessionId);
-          Logger.debug(`active session from client ${client.id}: ${session?.id}`);
+        if (isAuthV2Request(this, sessionToken) || isAuthV2Request(this, headerToken)) {
+          // Set Clerk session claims on request
+          // TBD Set on state / locals instead?
```
Comment (agis) [inline src/Clerk.ts:228]: ❓ What is this about? Is it blocking the merge of this PR?

File: src/Clerk.ts [hunk 5]
```diff
@@ -204,36 +218,45 @@ export default class Clerk {
...
+        Logger.debug(`headerToken: ${headerToken}`);
 
-        Logger.debug(`sessionId from query: ${sessionId}`);
-
-        let session: (Session | undefined) = undefined;
-        if (!sessionId) {
-          const client = await this.clients.verifyClient(sessionToken);
-          session = client.sessions.find(session => session.id === client.lastActiveSessionId);
-          Logger.debug(`active session from client ${client.id}: ${session?.id}`);
+        if (isAuthV2Request(this, sessionToken) || isAuthV2Request(this, headerToken)) {
```
Comment (agis) [inline src/Clerk.ts:226]: 🔧 Not that it should make any difference, but consider switching the order for consistency with the Ruby SDK.

File: src/Clerk.ts (hunk 3 above)
Comment (SokratisVidros) [reply src/Clerk.ts:208]: Correct

File: src/Clerk.ts [hunk 6]
```diff
@@ -204,36 +223,63 @@ export default class Clerk {
...
         }
 
-        let sessionId: any = req.query._clerk_session_id;
+        if (isAuthV2Request(this, headerToken) || isAuthV2Request(this, sessionToken)) {
+          // Set Clerk session claims on request
+          // TBD Set on state / locals instead?
+
+          try {
+            // @ts-ignore
+            req.sessionClaims = await this.verifyToken(headerToken);
```
Comment (SokratisVidros) [inline src/Clerk.ts:239]: Maybe rename this to `currentSID`? How do we do this across middlewares? The name of the key on the request object should be the same across ruby, python, go.

File: src/Clerk.ts [hunk 7]
```diff
@@ -204,36 +223,63 @@ export default class Clerk {
...
+          } catch(e) {
+            Logger.debug(e)
+          }
+
+          // @ts-ignore
+          if (!req.sessionClaims) {
+            try {
+              // @ts-ignore
+              req.sessionClaims = await this.verifyToken(sessionToken);
```
Comment (SokratisVidros) [inline src/Clerk.ts:248]:
```suggestion
              req.sessionClaims = await this.verifyToken(cookieToken);
```

File: src/Clerk.ts (hunk 6 above)
Comment (chanioxaris) [reply src/Clerk.ts:239]: I came up with this name but i don't think is the best. Open to any naming suggestion. Note that this should be different that the already session field, as at least in Go sdk we can't assign a different type in the same context key without being an `interface{}`, which will confuse the users

Comment (SokratisVidros) [reply src/Clerk.ts:239]: Let's go with currentSID then for now. I don't have a better suggestion. @colinclerk Any ideas?

File: src/Clerk.ts [hunk 8]
```diff
@@ -194,6 +200,27 @@ export default class Clerk {
...
+        const decoded = clerk.decodeToken(token)
+        return (decoded.iss?.lastIndexOf('https://clerk.', 0) === 0)
+      } catch (e) {
+        return false
+      }
+    }
+
+    async function verifyToken(clerk: Clerk, token: string):Promise<JwtPayload | undefined> {
+      try {
+        return await clerk.verifyToken(token)
```
Comment (SokratisVidros) [inline src/Clerk.ts:218]:
```suggestion
        return clerk.verifyToken(token)
```

File: src/Clerk.ts [hunk 9]
```diff
@@ -194,6 +200,27 @@ export default class Clerk {
...
+        }
+
+        const decoded = clerk.decodeToken(token)
+        return (decoded.iss?.lastIndexOf('https://clerk.', 0) === 0)
+      } catch (e) {
+        return false
+      }
+    }
+
+    async function verifyToken(clerk: Clerk, token: string):Promise<JwtPayload | undefined> {
```
Comment (SokratisVidros) [inline src/Clerk.ts:216]:
```suggestion
    function verifyToken(clerk: Clerk, token: string):Promise<JwtPayload | undefined> {
```

File: src/Clerk.ts [hunk 10]
```diff
@@ -203,37 +230,51 @@ export default class Clerk {
...
+        let headerToken;
+        if (req.headers) {
+          headerToken = (req.headers['Authorization'] || req.headers['authorization']) as string
+          Logger.debug(`headerToken: ${headerToken}`);
         }
 
-        let sessionId: any = req.query._clerk_session_id;
+        if (isAuthV2Request(this, headerToken) || isAuthV2Request(this, cookieToken)) {
+          // Set Clerk session claims on request
+          // TBD Set on state / locals instead?
```
Comment (SokratisVidros) [inline src/Clerk.ts:244]: It's ok to set it on the request for now.

**Review by SokratisVidros — APPROVED** [review]

**Pull request merged**